src/box/sql/sqlInt.h

```diff
@@ -834,9 +834,9 @@ typedef u64 uptr;
  * pointers.  In that case, only verify 4-byte alignment.
  */
 #ifdef SQL_4_BYTE_ALIGNED_MALLOC
-#define EIGHT_BYTE_ALIGNMENT(X)   ((((char*)(X) - (char*)0)&3)==0)
+#define EIGHT_BYTE_ALIGNMENT(X)   ((((uintptr_t)((char*)(X)) - (uintptr_t)((char*)0))&3)==0)
 #else
-#define EIGHT_BYTE_ALIGNMENT(X)   ((((char*)(X) - (char*)0)&7)==0)
+#define EIGHT_BYTE_ALIGNMENT(X)   ((((uintptr_t)((char*)(X)) - (uintptr_t)((char*)0))&7)==0)
```

Comment:
```suggestion
#define EIGHT_BYTE_ALIGNMENT(X)   ((((uintptr_t)((char*)(X)) & 7) == 0)
```

Reply:
Honestly, at first I also wondered about the reason to subtract the NULL address in the original macros. But it seems that some machines can have non-zero address of the NULL pointer.

Reply:
So, if you think that we never execute Tarantool on such strange machines, I'll remove subtraction of a pointer with a zero address.

Reply:
> But it seems that some machines can have non-zero address of the NULL pointer.

Could you provide a reference please?

Reply:
> So, if you think that we never execute Tarantool on such strange machines, I'll remove subtraction of a pointer with a zero address.
According to our [README](https://github.com/tarantool/tarantool/blob/master/README.md):
https://github.com/tarantool/tarantool/blob/7aa02f302d819727f68e8f455ea23a066a00b4b5/README.md?plain=1#L41-L42

Reply:
AFAIC, this subtraction was needed for an implicit conversion to [`ptrdiff_t`](https://en.cppreference.com/w/c/types/ptrdiff_t) without including the `stddef.h` header file.

Reply:
> Furthermore, the macros subtracted literal 0, not NULL

If I understand the non-zero NULL pointer architectures correctly (I have never seen them in reality, only googled), this statement is incorrect. The macros subtracted an address on the null pointer as `(char*)0` that returns `x0` on regular machines and some non-zero value on non-zeroed machines. It is compiler's responsibility to understand that `0` is a NULL pointer and translate it correctly to non-zero bytes.

Reply:
> So, I believe this macro has been inherited from the SQLite that decided to double check such an exotic case. It compares the distance in bytes between our address and the address of the NULL pointer (that can be non-zero value).

I really doubt this was the semantics that the authors implied. Judging by the name of the macro, I believe the purpose of this subtraction is the following:
> AFAIC, this subtraction was needed for an implicit conversion to [ptrdiff_t](https://en.cppreference.com/w/c/types/ptrdiff_t) without including the stddef.h header file.

Reply:
As far as I can trace back the source code history, the purpose of this code was really only to check alignment — not any kind of 'distance comparison' that you mentioned:
https://github.com/sqlite/sqlite/commit/c96d8530ca54c7ec7c6ca52b9e7f01f33f0c4235

https://github.com/sqlite/sqlite/commit/66e8008404d70107795683f0fa2b533e0e641f46